File: ts/src/rpc/streamHello.ts
import type { StreamHello } from "../gen/flowersec/rpc/v1.gen.js";
import { readJsonFrame, writeJsonFrame } from "./framing.js";

export async function writeStreamHello(write: (b: Uint8Array) => Promise<void>, kind: string): Promise<void> {
  const h: StreamHello = { kind, v: 1 };
  await writeJsonFrame(write, h);
}

export async function readStreamHello(readExactly: (n: number) => Promise<Uint8Array>): Promise<StreamHello> {
  const h = (await readJsonFrame(readExactly, 8 * 1024)) as StreamHello;
  if (h.v !== 1 || typeof h.kind !== "string" || h.kind.length === 0) throw new Error("bad StreamHello");
  return h;
}

File: go/rpc/stream_hello.go
package rpc

import (
	"encoding/json"
	"errors"
	"io"

	rpcv1 "github.com/flowersec/flowersec/gen/flowersec/rpc/v1"
)

var ErrBadStreamHello = errors.New("bad stream hello")

func WriteStreamHello(w io.Writer, kind string) error {
	return WriteJSONFrame(w, rpcv1.StreamHello{Kind: kind, V: 1})
}

func ReadStreamHello(r io.Reader, maxLen int) (rpcv1.StreamHello, error) {
	b, err := ReadJSONFrame(r, maxLen)
	if err != nil {
		return rpcv1.StreamHello{}, err
	}
	var h rpcv1.StreamHello
	if err := json.Unmarshal(b, &h); err != nil {
		return rpcv1.StreamHello{}, ErrBadStreamHello
	}
	if h.V != 1 || h.Kind == "" {
		return rpcv1.StreamHello{}, ErrBadStreamHello
	}
	return h, nil
}
